app/services/duplicate_profile_resolver.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+# A service class to merge duplicate Profile objects
+class DuplicateProfileResolver
+  class << self
+    def run!
+      @profiles = nil
+      duplicate_profiles.find_each do |profile|
+        if existing_profile(profile)
+          migrate_profile(existing_profile(profile), profile)
+        else
+          self.existing_profile = profile
+        end
+      end
+    end
+
+    private
+
+    def existing_profile(profile)
+      profiles[[profile.ref_id, profile.account_id, profile.benchmark_id]]
+    end
+
+    def existing_profile=(profile)
+      profiles[[profile.ref_id,
+                profile.account_id,
+                profile.benchmark_id]] = profile
+    end
+
+    def profiles
+      @profiles ||= {}
+    end
+
+    def migrate_profile(existing_p, duplicate_p)
+      migrate_profile_hosts(existing_p, duplicate_p)
+      migrate_test_results(existing_p, duplicate_p)
```

Comment:
You can skip these 2 if `profile.canonical`

Reply:
Well, at first I agreed with you for the sake of optimization, but these methods are already highly efficient, only generating a single SQL query; adding an additional check seems like premature optimization. IMO the simplicity here is very important to ensure correctness - one of the mistakes I made when first implementing the duplicate benchmark resolver.

I guess, one last comment: canonical profiles _shouldn't_ have profile_hosts or test_results, but why would we check `canonical` and rely on that assumption instead of checking for `profile_hosts` or `test_results` directly?

Reply:
I'd say if that happens it should straightaway break the migration to log it - no hosts or test results should ever be assigned to these